rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryInstances.scala

```diff
@@ -203,22 +203,115 @@ object HistoryInstances {
       FlatMap[F].tailRecM((fullPath, trieNodesOnPath, none[Trie]))(go)
     }
 
-    def process(actions: List[HistoryAction]): F[History[F]] =
-      // TODO this is an intermediate step to reproduce all the trie behavior
-      // will evolve to a fold based implementation with partial tries
+    def findUncommonNode(
+        currentPath: KeyPath,
+        previousPath: KeyPath,
+        previousRoot: Trie
+    ): F[Option[Blake2b256Hash]] = {
+      def commonPrefixWithLastDiverging(path: KeyPath, affix: ByteVector): Boolean =
+        affix.toSeq.startsWith(path.dropRight(1))
+
+      def handlePointerToPointerBlock(
+          ptr: ValuePointer
+      ): F[Either[(KeyPath, Trie), Option[Blake2b256Hash]]] =
+        ptr match {
+          case _: LeafPointer => Applicative[F].pure(none.asRight) // ignore leaf
+          case NodePointer(hash) =>
+            Applicative[F].pure(hash.some.asRight) // pick pointerBlock
+        }
+
+      def pathTerminatesInAffix(path: KeyPath, affix: KeyPath): Boolean =
+        affix == path || affix.startsWith(path) || (path.size == 1 && affix.startsWith(path))
+
+      type Params = (KeyPath, Trie)
```

Comment:
`Params` is so vague that I'd get rid of it

Reply:
changed to `PathToTrie`